Add the gem for user authentication with Google

kinyaoo opened the pull request:
#### Summary

Users should log in with their Google account. Adds the OmniAuth Google
OAuth2 gem in order to authenticate with Google.

#### Ticket Link

Github: https://github.com/mickmister/music-sniper-rails/issues/1
Gitlab: https://gitlab.com/mjkochell/music-sniper/issues/101

kinyaoo committed:
Add the gem for user authentication with Google

Users should log in with their Google account. Adds the OmniAuth Google
OAuth2 gem in order to authenticate with Google.

Implement: #1

mickmister commented:
Looking really good! I like the improvements added. Let's continue discussion here so it's closer to the code.

kinyaoo committed:
Extract user auth validation to the controller

The command which returns authentication information also verifies if
the auth data are valid with a specific user. Extracts the behavior to
the controller to permit to have more auth types that use the auth user
command.

Implement: #1

kinyaoo committed:
Implement Google Omniauth to authenticate users

Users need to authenticate through google oauth2, implements the
configuration to permit this auth.

Implement: #1

mickmister commented on app/commands/authenticate_user.rb:
```diff
@@ -1,31 +1,29 @@
 
   def call
-    if user
-      auth_token = JsonWebToken.encode(user_id: user.id) if user
```
This code was terrible! :laughing:

mickmister commented on app/commands/authenticate_user.rb:
```diff
@@ -1,31 +1,29 @@
+  end
+
+  def valid_user_auth?
+    return user && user.authenticate(password) if password
```
Minor suggestion

```suggestion
	return false unless user && password
	
	return user.authenticate(password)
```

mickmister replied:
Actually this doesn't work for empty strings. First line should be:

```
	return false unless [user, password].all?
```

mickmister reviewed (commented):
Great work @kinyaoo! Just one suggestion

Also, Github can lose track of comments when the original commit hash of the comment is changed. Do you mind pushing to your branch when you make updates, rather than rebasing and force-pushing?

kinyaoo commented:
> Great work @kinyaoo! Just one suggestion
> 
> Also, Github can lose track of comments when the original commit hash of the comment is changed. Do you mind pushing to your branch when you make updates, rather than rebasing and force-pushing?

I could it's just because I use to reorganize my commits by rebasing then I have to push force but commits are cleaner this way I think

mickmister commented:
> I could it's just because I use to reorganize my commits by rebasing then I have to push force but commits are cleaner this way I think

@kinyaoo Sounds good. Sorry about the lack of activity on my part. I will test sometime this week and get this merged.

mickmister commented:
@kinyaoo I wanted to see the feasibility of a client-side solution, and it turned out to be very few lines of code with https://github.com/anthonyjgrove/react-google-login.

1. The frontend authenticates with Google
2. The frontend sends to the backend a token that Google returned
3. The backend uses the token to get the user's Google-verified email address
4. The user now has an account, and there are no refresh tokens to be stored in the database

For the original purposes of this feature, these steps satisfy the requirement. I think your solution has the potential to provide many more features. Do you know what use cases the solution in this PR provides outside of the requirement?

kinyaoo commented:
@mickmister 

I haven't tried but according to https://github.com/omniauth/omniauth I think that this PR can be used to authenticate users through multiple strategies.

We just need to follow these steps :

1. Add and configure the corresponding gem from https://github.com/omniauth/omniauth/wiki/List-of-Strategies

2. Add the new provider in the config/initializers/omniauth.rb file

You can merge this if it's good for you and assign me to the next issues for implementing new providers for authentication if you want

kinyaoo commented:
Hi, could you review the PR ? @mickmister

mickmister commented on app/commands/authenticate_user.rb:
```diff
@@ -1,31 +1,25 @@
+    return invalid_credentials if @valid_user_auth.blank?
+
+    auth_token = JsonWebToken.encode(user_id: user.id)
+    { auth_token: auth_token, user: user.safe_attributes }
```
I have gotten rid of `user.safe_atrributes` in my code that I haven't merged yet. Can you please change this to `as_json` to match the changes I will introduce later?
```suggestion
    { auth_token: auth_token, user: user.as_json }
```

mickmister commented on test/models/o_auth_access_data_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class OAuthAccessDataTest < ActiveSupport::TestCase
```
Could you please write a test here or delete the file? What do you think you would want to / be able to test here?

kinyaoo replied:
I prefer to delete it for now

mickmister reviewed (changes requested):
Hi @kinyaoo, I have a few questions

Can you tell me about the lifecycle of a o_auth_access_data record? Do we need to refresh the tokens on some unit of time basis? Are they used upon re-login?

Can you provide steps for me to test this? Maybe something that you have done that I can replicate using a UI component?

Also I want to point out that if you rebase, I have to review all of the changes at once again. I understand that you want to reorganize the commits. But if you instead just add more commits onto the PR, it is easier for me to review as I can see the commits that I have not reviewed yet. That would be very helpful!

Thank you!

kinyaoo committed:
Convert user as json during auth

kinyaoo commented:
> Hi @kinyaoo, I have a few questions
> 
> Can you tell me about the lifecycle of a o_auth_access_data record? Do we need to refresh the tokens on some unit of time basis? Are they used upon re-login?

Refresh token will never change there is only the access token every 2 hours I think. You can easily re-login your user.

> Can you provide steps for me to test this? Maybe something that you have done that I can replicate using a UI component?

Using a UI component you juste have to go on the following URL http://localhost:3000/auth/google_oauth2

> Also I want to point out that if you rebase, I have to review all of the changes at once again. I understand that you want to reorganize the commits. But if you instead just add more commits onto the PR, it is easier for me to review as I can see the commits that I have not reviewed yet. That would be very helpful!

Ok I understand

mickmister committed:
Merge remote-tracking branch 'origin' into feature/google_oauth-merge

mickmister commented on app/models/update_o_auth_access_data_form.rb:
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+class UpdateOAuthAccessDataForm
```
Can you help me understand why this code has its own model, as opposed to living in the `OAuthAccessData` class, or something else e.g. a service?

kinyaoo replied:
This article will explain better than me why I use form objects.
But I agree with you it can be confusing to let this code in the models directory.
We could move this code under a `forms` directory if you prefer ?

mickmister replied:
> We could move this code under a forms directory if you prefer?

Sounds good to me!

mickmister committed:
remove foreign key

mickmister commented on app/controllers/omniauth_sessions_controller.rb:
```diff
@@ -0,0 +1,37 @@
+    else
+      unauthorized(errors: command.errors)
+    end
+  end
```
Can we return a useful error instead of the two json responses here? This is shown as raw data in the user's browser after they authenticate. Maybe something like `Connected (email) account` and `Error authenticating`.

kinyaoo replied:
Something like :
```
if command success?
 render json: "Connected #{command[:user][:email] account"
else
 unauthorized(errors: "Error authenticating")
end
```

mickmister replied:
Instead of json, I think we should return html or text.

```ruby
if command.success?
  render html: "Connected #{command.result[:user]['email']} account. You may now close this browser tab."
else
  render html: "Unable to connect account."
end
```

You will also need a `skip_before_action :force_json` at the top of the file to be able to use html.

mickmister replied:
Feel free to use some sort of html here if you have a more elegant idea! 

EDIT: If you do decide to use html, I would keep it simple so you can keep it inline here, or move it to a view.

mickmister reviewed (changes requested):
I've gotten this PR up to speed with master. I've confirmed the feature works correctly. Great job @kinyaoo! :tada:

I've left two comments for feedback. Once they are addressed, we will merge this into master.

kinyaoo commented:
> I've gotten this PR up to speed with master. I've confirmed the feature works correctly. Great job @kinyaoo! 

Thank you it's a pleasure @mickmister

kinyaoo committed:
Render HTML instead of json with Omniauth session

Implement: #1

kinyaoo commented:
@mickmister I just pushed the last commit, sorry for the delay

mickmister reviewed (approved):
Awesome work!! Thanks for addressing all the feedback @kinyaoo
